bibtex-actions.el

```diff
@@ -254,7 +256,7 @@ offering the selection candidates"
     (seq-map
      (lambda (choice)
        ;; Collect citation keys of selected candidate(s).
-       (or (cadr (assoc choice candidates))
+       (or (cdr (assoc choice candidates))
            ;; Key is literal coming from embark, just pass it on
            choice))
```

Comment:
This change is going to break embark as it is, since now the functions expect an alist of keys and entries and not a list of keys. So the `choice` needs to be changed to something like
```elisp
(cdr (seq-find (lambda (cand) (equal choice (cadr cand))) candidates))
```
to complete the key at point into a key entry `cons`.

Reply:
> This change is going to break embark as it is ...

Yeah, I was aware of that; was going to figure that out next, so thanks!

Reply:
> So the `choice` needs to be changed to something like
> 
> ```emacs-lisp
> (cdr (seq-find (lambda (cand) (equal choice (cadr cand))) candidates))
> ```

Actually, I had been thinking the functions in the "Embark" section would have to be changed; not this?

Reply:
> Actually, I had been thinking the functions in the "Embark" section would have to be changed; not this?

According to what Omar said last time, emark will basically call the commands interactively and that in turn means they will call `select-keys` and embark will simulate typing the keys at point. Here we couple the passed key with the corresponding entry. I think this should work for embark and I don't see anywhere else we can do this coupling

Reply:
> If I'm right, I'm not even sure how to do that ATM, as we only have access to the keys in the buffer.

The only difference between the normal usage of `select-keys` and embark is that in the first case the returned strings are the display strings and in the case of embark they are keys. But we can use either to find the key entry pair, so they are not that different from that perspective.

Reply:
But embark is expecting a string.

```elisp
Error running timer: (wrong-type-argument stringp ("agnew2003" ("title" . "From Megalopolis to Global City-Region?") ("date" . "2003") ("author" . "Agnew, John") ("=type=" . "article") ("=key=" . "agnew2003")))
```

Reply:
> But embark is expecting a string.
> 
> ```emacs-lisp
> Error running timer: (wrong-type-argument stringp ("agnew2003" ("title" . "From Megalopolis to Global City-Region?") ("date" . "2003") ("author" . "Agnew, John") ("=type=" . "article") ("=key=" . "agnew2003")))
> ```

Embark should get a string at least that is my impression after reading Omar's reply again. Though I would expect this error with `default-action` because of the string splitting going on in `run-default-action`. Specifically due to `(split-string (cdr keys) " & ")` which is trying to string-split an alist as in the error. I don't think any processing of input in needed. there.

Reply:
I get this error using embark in the minibuffer.l.

I'll probably need to look at this again tomorrow, with fresher eyes.

The embark stuff has proved kind of brittle.

Reply:
> I get this error using embark in the minibuffer.l.

The at point stuff works then? I was thinking only of that.

I don't know how embark works in the minibuffer, but that error looks like one of our commands is trying to string-split an alist and might be unrelated to embark.

Reply:
No, I hadn't even gotten to the at-point stuff. Now that I haven't that working, I'll try this again.

> ... that error looks like one of our commands is trying to string-split an alist and might be unrelated to embark.

Yes. I had not yet modified the notes function to take key-entry rather than key.

That one works now, but that introduces a design question:

I designed that function so one could swap in others, including the edit-note function from org-roam-bibtex. But that only takes a key.

Reply:
> No, I hadn't even gotten to the at-point stuff. Now that I haven't that working, I'll try this again.

Looks like your suggestion work, though I still don't understand it :-)

Reply:
Are you comfortable with me pulling the trigger on this generally, once I have it sorted out?

E.g. changing the interactive functions so dramatically, and diverging clearly from other similar packages?

Reply:
> Based on initial test, looks like your suggestion works, though I still don't understand why :-)

The details are in https://github.com/bdarcus/bibtex-actions/issues/233#issuecomment-901536901

But briefly,
1) Embark calls some command interactively
2) This means getting the alist of `(key . entry)` by calling `select-keys`
3) This calls `completing-read-multiple`
4) Embark then "types" `"key1 & key2 & .. & keyn"` in the minibuffer and presses RET.
5) completing-read-multiple then hands us a list of keys `(list "key1" "key2" ... "keyn")`
6) My suggestion completes the list of keys into the format we want

Reply:
I want to put a clear comment there, to replace the earlier one you had; does this work?

```elisp
           ;; When calling embark at-point, use keys to look up and return the
           ;; selected candidates.
```

BTW, I changed `select-keys` to `select-refs`.

Reply:
> That one works now, but that introduces a design question:
> 
> I designed that function so one could swap in others, including the edit-note function from org-roam-bibtex. But that only takes a key.

I think the point of the change is to restrict the use of cache so that it happens only through `select-keys`. Other commands can restrict themselves to suitable parts of the bibtex-entry, especially if that involves going through a `defcustom`. I don't understand the notes function but I think two options are 
1) There is no need for the notes function to consider itself with the details of the entry so in that case we use the `extract-keys` internally and expose only the `key` to be used as an input.

2) The rest of the information might be useful but the default using `org-roam-bibtex` doesn't need it. In that case the keys should be extracted in the definition of the default for corresponding `defcustom`.

Reply:
> Are you comfortable with me pulling the trigger on this generally, once I have it sorted out?
> 
> E.g. changing the interactive functions so dramatically, and diverging clearly from other similar packages?
> 
> I think I am, but just want a sanity check!

I think overall it is a change for the better to make cache more of an internal implementation detail to be used sparingly. But it probably will break setup of some people with customization in their config. Might be a good idea to make an issue and allow people to chime in. 

But I am fairly trigger happy with changes and I think we have changed internals a few times since I started working on it without too many problems.

Reply:
> I want to put a clear comment there, to replace the earlier one you had; does this work?
> 
> ```emacs-lisp
>            ;; When calling embark at-point, use keys to look up and return the
>            ;; selected candidates.
> ```

This seems better. Might be a good idea to link to Omar's reply there so if someone new comes along, they can pick up how it is working quickly. It cleared up a few things for me.

> BTW, I changed `select-keys` to `select-refs`.

Seems good!

Reply:
> I don't understand the notes function but I think two options are
> 
> 1. There is no need for the notes function to consider itself with the details of the entry so in that case we use the `extract-keys` internally and expose only the `key` to be used as an input.

The function does need to concern itself with the "details of the entry" because it has to run format-entry on the note template for new notes.

But ...

> 2. The rest of the information might be useful but the default using org-roam-bibtex doesn't need it. 

... this is also true, I guess, and less-than-ideal, since ORB relies on the bibtex-completion cache.

Perhaps I could ask them to allow an optional "entry" arg.

Reply:
> The function does need to concern itself with the "details of the entry" because it has to run format-entry on the note template for new notes.

Then I think this change should be better since now the function is given entry upfront instead of having to find it.

> ... this is also true, I guess, and less-than-ideal, since ORB relies on the bibtex-completion cache; why they can get away with only using the key.

This is indeed unfortunate and I think a potential problem in removing bibtex-completion as a dependency since it will get loaded as a transitive one. 
 
> Perhaps I could ask them to allow an optional "entry" arg.

This seems like a good idea especially in view of minimizing dependencies.

>In any case, given status quo, I'm faced with a somewhat difficult choice:
>
>    do the "right" thing from a technical POV
>   do the "compatible" thing

> Or I could do a hybrid; take key, but get the entry from the cache.

I think for now the best option seems to be to have the customizable function take `(key . entry)`, but extract the key in the default and pass it to ORB and hope that they change to implementation so we can pass the entry directly in the future.

Reply:
> The function does need to concern itself with the "details of the entry" because it has to run format-entry on the note template for new notes.

Then I think this change should be better since now the function is given entry upfront instead of having to find it.

> ... this is also true, I guess, and less-than-ideal, since ORB relies on the bibtex-completion cache; why they can get away with only using the key.

This is indeed unfortunate and I think a potential problem in removing bibtex-completion as a dependency since it will get loaded as a transitive one. 
 
> Perhaps I could ask them to allow an optional "entry" arg.

This seems like a good idea especially in view of minimizing dependencies.

>In any case, given status quo, I'm faced with a somewhat difficult choice:
>
>    do the "right" thing from a technical POV
>   do the "compatible" thing

> Or I could do a hybrid; take key, but get the entry from the cache.

I think for now the best option seems to be to have the customizable function take `(key . entry)`, but extract the key in the default and pass it to ORB and hope that they change to implementation so we can pass the entry directly in the future.

Reply:
> I think for now the best option seems to be to have the customizable function take `(key . entry)`, but extract the key in the default and pass it to ORB and hope that they change to implementation so we can pass the entry directly in the future.

I pinged the lead maintainer, and he's open to it.

https://github.com/bdarcus/bibtex-actions/discussions/270#discussioncomment-1314505